cni/pkg/install/install.go

```diff
@@ -48,35 +48,50 @@ func NewInstaller(cfg *config.InstallConfig, isReady *atomic.Value) *Installer {
 	}
 }
 
+func (in *Installer) install(ctx context.Context) (err error) {
+	if err = copyBinaries(
+		in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs,
+		in.cfg.UpdateCNIBinaries, in.cfg.SkipCNIBinaries); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCopyBinariesFailure)).Increment()
+		return
+	}
+
+	if in.saToken, err = readServiceAccountToken(); err != nil {
+		cniInstalls.With(resultLabel.Value(resultReadSAFailure)).Increment()
+		return
+	}
+
+	if in.kubeconfigFilepath, err = createKubeconfigFile(in.cfg, in.saToken); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCreateKubeConfigFailure)).Increment()
+		return
+	}
+
+	if in.cniConfigFilepath, err = createCNIConfigFile(ctx, in.cfg, in.saToken); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCreateCNIConfigFailure)).Increment()
+		return
+	}
+
+	return
+}
+
 // Run starts the installation process, verifies the configuration, then sleeps.
 // If an invalid configuration is detected, the installation process will restart to restore a valid state.
 func (in *Installer) Run(ctx context.Context) (err error) {
-	for {
-		if err = copyBinaries(
-			in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs,
-			in.cfg.UpdateCNIBinaries, in.cfg.SkipCNIBinaries); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCopyBinariesFailure)).Increment()
-			return
-		}
-
-		if in.saToken, err = readServiceAccountToken(); err != nil {
-			cniInstalls.With(resultLabel.Value(resultReadSAFailure)).Increment()
-			return
-		}
+	if err = in.install(ctx); err != nil {
+		return
+	}
 
-		if in.kubeconfigFilepath, err = createKubeconfigFile(in.cfg, in.saToken); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCreateKubeConfigFailure)).Increment()
+	for {
+		if err = sleepCheckInstall(ctx, in.cfg, in.cniConfigFilepath, in.isReady); err != nil {
 			return
 		}
 
-		if in.cniConfigFilepath, err = createCNIConfigFile(ctx, in.cfg, in.saToken); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCreateCNIConfigFailure)).Increment()
-			return
+		if in.cfg.CNIEnableReinstall {
+			if err = in.install(ctx); err != nil {
+				return
+			}
 		}
 
-		if err = sleepCheckInstall(ctx, in.cfg, in.cniConfigFilepath, in.isReady); err != nil {
-			return
-		}
 		// Invalid config; pod set to "NotReady"
 		installLog.Info("Restarting...")
```

Comment:
Let's make the log more clear. We need to move this to line 88, and let's change it to be more informative (`Detect changes to the CNI configuration and binaries, attempt reinstalling...`). Then add an `else` block at line 93 with a log line saying (`Skip reinstalling CNI configuration and binaries.`)